Import initial generated 1.85 relnotes (#135797)
rust-lang/rust: RELEASES.md

  See [RFC 3668](https://rust-lang.github.io/rfcs/3668-async-closures.html) for more details.
- [Stabilize `#[diagnostic::do_not_recommend]`](https://github.com/rust-lang/rust/pull/132056)
- [Add `unpredictable_function_pointer_comparisons` lint to warn against function pointer comparisons](https://github.com/rust-lang/rust/pull/118833)
- [Lint on combining `#[no_mangle]` and `#[export_name]` attributes.](https://github.com/rust-lang/rust/pull/131558)

<a id="1.85.0-Compiler"></a>

Compiler
--------
- [The unstable flag `-Zpolymorphize` has been removed](https://github.com/rust-lang/rust/pull/133883), see https://github.com/rust-lang/compiler-team/issues/810 for some background.

<a id="1.85.0-Platform-Support"></a>

Platform Support
----------------
- [Promote `powerpc64le-unknown-linux-musl` to tier 2 with host tools](https://github.com/rust-lang/rust/pull/133801)

Refer to Rust's [platform support page][platform-support-doc]
for more information on Rust's tiered platform support.

<a id="1.85.0-Libraries"></a>

Libraries
---------
- [Panics in the standard library now have a leading `library/` in their path](https://github.com/rust-lang/rust/pull/132390)

> - [`std::env::home_dir()` on Windows now ignores the non-standard `$HOME` environment variable](https://github.com/rust-lang/rust/pull/132515)
>    It will be un-deprecated in a subsequent release.

Reviewer: ```suggestion
- [`std::env::home_dir()` on Windows now ignores the non-standard `$HOME` environment variable](https://github.com/rust-lang/rust/pull/132515)

  It will be un-deprecated in a subsequent release.
```

- [`<float>::min`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.min)
- [`<float>::clamp`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.clamp)
- [`<float>::abs`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.abs)
- [`<float>::signum`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.signum)
- [`<float>::copysign`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.copysign)
- [`MaybeUninit::write`](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.write)

<a id="1.85.0-Cargo"></a>

Cargo
-----
- [Add future-incompatibility warning against keywords in cfgs and add raw-idents](https://github.com/rust-lang/cargo/pull/14671/)
- [Stabilize higher precedence trailing flags](https://github.com/rust-lang/cargo/pull/14900/)
- [Pass `CARGO_CFG_FEATURE` to build scripts](https://github.com/rust-lang/cargo/pull/14902/)

<a id="1.85.0-Rustdoc"></a>

Rustdoc
-----
- [Doc comment on impl blocks shows the first line, even when the impl block is collapsed](https://github.com/rust-lang/rust/pull/132155)

<a id="1.85.0-Compatibility-Notes"></a>

Compatibility Notes
-------------------

> - [`rustc` no longer treats the `test` cfg as a well known check-cfg](https://github.com/rust-lang/rust/pull/131729), instead it is up to the build systems and users of `--check-cfg`[^check-cfg] to set it as a well known cfg using `--check-cfg=cfg(test)`.
>   This is done to enable build systems like Cargo to set it conditionally, as not all source files are suitable for unit tests.
>   [Cargo (for now) unconditionally sets the `test` cfg as a well known cfg](https://github.com/rust-lang/cargo/pull/14963).
> [^check-cfg]: https://doc.rust-lang.org/nightly/rustc/check-cfg.html

Reviewer: ```suggestion
- [`rustc` no longer treats the `test` cfg as a well known check-cfg](https://github.com/rust-lang/rust/pull/131729), instead it is up to the build systems and users of `--check-cfg`[^check-cfg] to set it as a well known cfg using `--check-cfg=cfg(test)`.

  This is done to enable build systems like Cargo to set it conditionally, as not all source files are suitable for unit tests.
  [Cargo (for now) unconditionally sets the `test` cfg as a well known cfg](https://github.com/rust-lang/cargo/pull/14963).
  [^check-cfg]: https://doc.rust-lang.org/nightly/rustc/check-cfg.html
```

Reviewer: Looks like the squash you did @Mark-Simulacrum didn't take into account those empty lines we added, they are removed from your last commit and subsequently from what was merged in master.

Opened #137125 to fix it.

Author: Thanks! Probably forgot to pull :(

- [Disable potentially incorrect type inference if there are trivial and non-trivial where-clauses](https://github.com/rust-lang/rust/pull/132325)
- `std::env::home_dir()` has been deprecated for years, because it can give surprising results in some Windows configurations if the `HOME` environment variable is set (which is not the normal configuration on Windows). We had previously avoided changing its behavior, out of concern for compatibility with code depending on this non-standard configuration. Given how long this function has been deprecated, we're now fixing its behavior as a bugfix. A subsequent release will remove the deprecation for this function.

> - [Make `core::ffi::c_char` signedness more closely match that of the platform-default `char`](https://github.com/rust-lang/rust/pull/132975)
>   This changed `c_char` from an `i8` to `u8` or vice versa on many Tier 2 and 3
>   targets (mostly Arm and RISC-V embedded targets). The new definition may
>   result in compilation failures but fixes compatibility issues with C.
>   The `libc` crate matches this change as of its 0.2.169 release.

Reviewer: ```suggestion
- [Make `core::ffi::c_char` signedness more closely match that of the platform-default `char`](https://github.com/rust-lang/rust/pull/132975)

  This changed `c_char` from an `i8` to `u8` or vice versa on many Tier 2 and 3
  targets (mostly Arm and RISC-V embedded targets). The new definition may
  result in compilation failures but fixes compatibility issues with C.

  The `libc` crate matches this change as of its 0.2.169 release.
```

Version 1.85.0 (2025-02-20)
==========================

<a id="1.85.0-Language"></a>

Language
--------
- [Support var-args for more calling conventions](https://github.com/rust-lang/rust/pull/116161) 

>   This stabilizes usage of `...` variable args in extern declarations for `cdecl`, `system`, `aapcs`, `sysv64`, `win64`, and `efiapi` calling conventions.

Reviewer 2: ```suggestion
  This stabilizes usage of `...` variable args in extern declarations for `C`, `cdecl`, `system`, `aapcs`, `sysv64`, `win64`, and `efiapi` calling conventions.
```

Reviewer 3: Weren't `C` and `cdecl` already stable?

- [Add `unpredictable_function_pointer_comparisons` lint to warn against function pointer comparisons](https://github.com/rust-lang/rust/pull/118833)
- [Lint on combining `#[no_mangle]` and `#[export_name]` attributes.](https://github.com/rust-lang/rust/pull/131558)
- [Stabilize `#[diagnostic::do_not_recommend]`](https://github.com/rust-lang/rust/pull/132056)

> - [Stabilize async closures (RFC 3668)](https://github.com/rust-lang/rust/pull/132706)

Reviewer 2: ```suggestion
- [Stabilize `async_closure`](https://github.com/rust-lang/rust/pull/132706) See [RFC 3668](https://rust-lang.github.io/rfcs/3668-async-closures.html) for more details.
```

- [The 2024 Edition is now stable.](https://github.com/rust-lang/rust/pull/133349)
  See [the edition guide](https://doc.rust-lang.org/nightly/edition-guide/rust-2024/index.html) for more details.

<a id="1.85.0-Compiler"></a>

Compiler
--------
- [The unstable flag `-Zpolymorphize` has been removed](https://github.com/rust-lang/rust/pull/133883), see https://github.com/rust-lang/compiler-team/issues/810 for some background.

> - [Enable LSX feature for `loongarch64_unknown_linux_ohos`.](https://github.com/rust-lang/rust/pull/134751)

Reviewer 4: Suggestion: probably remove this (maybe by closing the relnotes issue?), I know I was the one who tagged that with relnotes, but I think for Tier 3 targets we usually don't list enabling a target feature?

<a id="1.85.0-Platform-Support"></a>

Platform Support
----------------
- [Promote `powerpc64le-unknown-linux-musl` to tier 2 with host tools](https://github.com/rust-lang/rust/pull/133801)

Refer to Rust's [platform support page][platform-support-doc]
for more information on Rust's tiered platform support.

<a id="1.85.0-Libraries"></a>

Libraries
---------
- [Panics in the standard library now have a leading `library/` in their path](https://github.com/rust-lang/rust/pull/132390)

> - [Fix and undeprecate home_dir()](https://github.com/rust-lang/rust/pull/132515)

Reviewer 4: Nit:

```suggestion
- [Fix and undeprecate `std::env::home_dir`](https://github.com/rust-lang/rust/pull/132515)
```

- [Add `AsyncFn*` to the prelude in all editions.](https://github.com/rust-lang/rust/pull/132611)

<a id="1.85.0-Stabilized-APIs"></a>

Stabilized APIs
---------------

- [`BuildHasherDefault::new`](https://doc.rust-lang.org/stable/std/hash/struct.BuildHasherDefault.html#method.new)
- [`ptr::fn_addr_eq`](https://doc.rust-lang.org/std/ptr/fn.fn_addr_eq.html)
- [`io::ErrorKind::QuotaExceeded`](https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.QuotaExceeded)
- [`io::ErrorKind::CrossesDevices`](https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.CrossesDevices)
- [`{float}::midpoint`](https://doc.rust-lang.org/core/primitive.f32.html#method.midpoint)
- [Unsigned `{integer}::midpoint`](https://doc.rust-lang.org/std/primitive.u64.html#method.midpoint)
- [`NonZeroU*::midpoint`](https://doc.rust-lang.org/std/num/type.NonZeroU32.html#method.midpoint)

> - [impl std::iter::Extend for tuples with arity 1 through 12](https://doc.rust-lang.org/stable/std/iter/trait.Extend.html#impl-Extend%3C(A,)%3E-for-(EA,))

Reviewer 4: Nit:
```suggestion
- [impl `std::iter::Extend` for tuples with arity 1 through 12](https://doc.rust-lang.org/stable/std/iter/trait.Extend.html#impl-Extend%3C(A,)%3E-for-(EA,))
```

> - [FromIterator<(A, ...)> for (A, ...)](https://doc.rust-lang.org/stable/std/iter/trait.FromIterator.html#impl-FromIterator%3C(EA,)%3E-for-(A,)) (FromIterator for N-tuple for N in 1..=12)

Reviewer 4: Nit:
```suggestion
- [`FromIterator<(A, ...)>` for tuples with arity 1 through 12](https://doc.rust-lang.org/stable/std/iter/trait.FromIterator.html#impl-FromIterator%3C(EA,)%3E-for-(A,)) (FromIterator for N-tuple for N in 1..=12)
```

- [`BuildHasherDefault::new`](https://doc.rust-lang.org/stable/std/hash/struct.BuildHasherDefault.html#method.new)
- [`<float>::recip`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.recip)
- [`<float>::to_degrees`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.to_degrees)
- [`<float>::to_radians`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.to_radians)
- [`<float>::max`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.max)
- [`<float>::min`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.min)
- [`<float>::clamp`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.clamp)
- [`<float>::abs`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.abs)
- [`<float>::signum`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.signum)
- [`<float>::copysign`](https://doc.rust-lang.org/stable/std/primitive.f32.html#method.copysign)
- [`MaybeUninit::write`](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.write)

<a id="1.85.0-Cargo"></a>

Cargo
-----

> - [Add future-incompat warning against keywords in cfgs and add raw-idents](https://github.com/rust-lang/cargo/pull/14671/)

Reviewer 4: Nit:

```suggestion
- [Add future-incompatibility warning against keywords in cfgs and add raw-idents](https://github.com/rust-lang/cargo/pull/14671/)
```

> - [fix(cargo-rustc): stabilize higher precedence trailing flags](https://github.com/rust-lang/cargo/pull/14900/)

Reviewer 4: Nit:

```suggestion
- [Stabilize higher precedence trailing flags](https://github.com/rust-lang/cargo/pull/14900/)
```

> - [feat(build-script): Pass CARGO_CFG_FEATURE ](https://github.com/rust-lang/cargo/pull/14902/)

Reviewer 4: Nit:

```suggestion
- [Pass `CARGO_CFG_FEATURE` to build scripts](https://github.com/rust-lang/cargo/pull/14902/)
```

<a id="1.85.0-Rustdoc"></a>

Rustdoc
-----
- [Doc comment on impl blocks shows the first line, even when the impl block is collapsed](https://github.com/rust-lang/rust/pull/132155)

<a id="1.85.0-Compatibility-Notes"></a>

Compatibility Notes
-------------------
- [`rustc` no longer treats the `test` cfg as a well known check-cfg](https://github.com/rust-lang/rust/pull/131729), instead it is up to the build systems and users of `--check-cfg`[^check-cfg] to set it as a well known cfg using `--check-cfg=cfg(test)`.
  This is done to enable build systems like Cargo to set it conditionally, as not all source files are suitable for unit tests.

>   [Cargo (for now) unconditionally sets the `test` cfg as a well known cfg](https://github.com/rust-lang/cargo/pull/14963).

Reviewer 2: ```suggestion
- [Cargo (for now) unconditionally sets the `test` cfg as a well known cfg](https://github.com/rust-lang/cargo/pull/14963).
```

Reviewer 3: I'm pretty sure this is meant to be a continuation of the previous lines, not its own bullet.

[^check-cfg]: https://doc.rust-lang.org/nightly/rustc/check-cfg.html
- [Disable potentially incorrect type inference if there are trivial and non-trivial where-clauses](https://github.com/rust-lang/rust/pull/132325)
- `std::env::home_dir()` has been deprecated for years, because it can give surprising results in some Windows configurations if the `HOME` environment variable is set (which is not the normal configuration on Windows). We had previously changed its behavior, out of concern for compatibility with code depending on this non-standard configuration. Given how long this function has been deprecated, we're now fixing its behavior as a bugfix, and the next version will undeprecate it. (We deferred undeprecating it on nightly until today's stable release, to avoid potential confusion between the fixed version on nightly and the previously unfixed version on stable.)

> - [More closely match clang behavior for definition of core::ffi::c_char](https://github.com/rust-lang/rust/pull/132975)

Reviewer 4: Nit:

```suggestion
- [More closely match clang behavior for definition of `core::ffi::c_char`](https://github.com/rust-lang/rust/pull/132975)
```

>   On many targets, this has changed the definition of c_char, resulting in potential compilation issues.

Reviewer 4: Nit:

```suggestion
  On many targets, this has changed the definition of `c_char`, resulting in potential compilation issues.
```

>   The new definition is more likely to be accurate for the corresponding C definition on the relevant target.

Reviewer 4: Question: does this actually mean the "platform default C definition"?

Author: I don't know what the actual change did, this is a best effort description. I don't mind clarifying, but it seems good to ask on the relevant tracking issue (https://github.com/rust-lang/rust/issues/134157) to centralize discussion and reduce noise if we ping folks to find out.

Reviewer 5: I think it may be important to mention that this this doesn't affect much than T2/T3 no_std so most people can dismiss it quickly. Updated the notes in https://github.com/rust-lang/rust/issues/134157 to mention that, and slightly clarify the above.

>   The libc crate matches this change as of its 0.2.169 release.

Reviewer 4: Nit:

```suggestion
  The `libc` crate matches this change as of its 0.2.169 release.
```

- [Increase `sparcv9-sun-solaris` and `x86_64-pc-solaris` Solaris baseline to 11.4.](https://github.com/rust-lang/rust/pull/133293)
- [Show `abi_unsupported_vector_types` lint in future breakage reports](https://github.com/rust-lang/rust/pull/133374)

> - [Error if multiple super-trait instantiations of `dyn Trait` need assoc but only one is provided](https://github.com/rust-lang/rust/pull/133392)

Reviewer 4: Nit:

```suggestion
- [Error if multiple super-trait instantiations of `dyn Trait` need associated types to be specified but only one is provided](https://github.com/rust-lang/rust/pull/133392)
```

- [Change `powerpc64-ibm-aix` default `codemodel` to large](https://github.com/rust-lang/rust/pull/133811)

<a id="1.85.0-Internal-Changes"></a>

Internal Changes
----------------

These changes do not affect any public interfaces of Rust, but they represent
significant improvements to the performance or internals of rustc and related
tools.

> - [Build x86_64-unknown-linux-gnu with LTO for C/C++ code (e.g., jemalloc)](https://github.com/rust-lang/rust/pull/134690)

Reviewer 4: Nit:

```suggestion
- [Build `x86_64-unknown-linux-gnu` with LTO for C/C++ code (e.g., `jemalloc`)](https://github.com/rust-lang/rust/pull/134690)
```

Version 1.85.0 (2025-02-20)
==========================

<a id="1.85.0-Language"></a>

Language
--------
- [Support var-args for more calling conventions](https://github.com/rust-lang/rust/pull/116161) 

>   This stabilizes usage of `...` variable args in extern declarations for `system`, `aapcs`, `sysv64`, `win64`, and `efiapi` calling conventions. `C` and `cdecl` were previously stable.

Reviewer 3: This was reverted in #136897, and pending beta backport in #136980.
(I'll close the relnotes issue too.)

- [Add `unpredictable_function_pointer_comparisons` lint to warn against function pointer comparisons](https://github.com/rust-lang/rust/pull/118833)
- [Lint on combining `#[no_mangle]` and `#[export_name]` attributes.](https://github.com/rust-lang/rust/pull/131558)

> - [Stabilize `#[diagnostic::do_not_recommend]`](https://github.com/rust-lang/rust/pull/132056)

Reviewer 6: This one is worth a brief mention in the blog post.

> - [Stabilize async closures](https://github.com/rust-lang/rust/pull/132706)

Reviewer 7: And should probably appear second in the list, after the edition.

  See [RFC 3668](https://rust-lang.github.io/rfcs/3668-async-closures.html) for more details.

> - [The 2024 Edition is now stable.](https://github.com/rust-lang/rust/pull/133349)

Reviewer 6: I think this should be moved to the top of the list.

  See [the edition guide](https://doc.rust-lang.org/nightly/edition-guide/rust-2024/index.html) for more details.

<a id="1.85.0-Compiler"></a>

Compiler
--------
- [The unstable flag `-Zpolymorphize` has been removed](https://github.com/rust-lang/rust/pull/133883), see https://github.com/rust-lang/compiler-team/issues/810 for some background.

<a id="1.85.0-Platform-Support"></a>

Platform Support
----------------
- [Promote `powerpc64le-unknown-linux-musl` to tier 2 with host tools](https://github.com/rust-lang/rust/pull/133801)

Refer to Rust's [platform support page][platform-support-doc]
for more information on Rust's tiered platform support.

<a id="1.85.0-Libraries"></a>

Libraries
---------
- [Panics in the standard library now have a leading `library/` in their path](https://github.com/rust-lang/rust/pull/132390)

> - [`std::env::home_dir()` no longer respects `$HOME` on Windows](https://github.com/rust-lang/rust/pull/132515)

Reviewer 6: ```suggestion
- [`std::env::home_dir()` on Windows now ignores the non-standard `$HOME` environment variable](https://github.com/rust-lang/rust/pull/132515)
```

   It will be un-deprecated in a subsequent release.
- [Add `AsyncFn*` to the prelude in all editions.](https://github.com/rust-lang/rust/pull/132611)

<a id="1.85.0-Stabilized-APIs"></a>

Stabilized APIs
---------------

- [`BuildHasherDefault::new`](https://doc.rust-lang.org/stable/std/hash/struct.BuildHasherDefault.html#method.new)
- [`ptr::fn_addr_eq`](https://doc.rust-lang.org/std/ptr/fn.fn_addr_eq.html)
- [`io::ErrorKind::QuotaExceeded`](https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.QuotaExceeded)
- [`io::ErrorKind::CrossesDevices`](https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.CrossesDevices)
- [`{float}::midpoint`](https://doc.rust-lang.org/core/primitive.f32.html#method.midpoint)
- [Unsigned `{integer}::midpoint`](https://doc.rust-lang.org/std/primitive.u64.html#method.midpoint)
- [`NonZeroU*::midpoint`](https://doc.rust-lang.org/std/num/type.NonZeroU32.html#method.midpoint)
- [impl `std::iter::Extend` for tuples with arity 1 through 12](https://doc.rust-lang.org/stable/std/iter/trait.Extend.html#impl-Extend%3C(A,)%3E-for-(EA,))

> - [`FromIterator<(A, ...)>` for tuples with arity 1 through 12](https://doc.rust-lang.org/stable/std/iter/trait.FromIterator.html#impl-FromIterator%3C(EA,)%3E-for-(A,))

Reviewer 8: These two lines about tuples in collect/extend should probably get a mention in a blog post; that’s effectively a new major feature from the perspective of people “just using” Rust iterators!
